aten/src/ATen/Allocator.h

```diff
@@ -3,14 +3,98 @@
 #include <memory>
 #include <stddef.h>
 
-#include "ATen/Retainable.h"
+#include <ATen/Error.h>
+#include <ATen/Retainable.h>
+#include <ATen/Device.h>
+#include <ATen/detail/UniqueVoidPtr.h>
 
 namespace at {
 
+// A DataPtr is a unique pointer (with an attached deleter and some
+// context for the deleter) to some memory, which also records what
+// device is for its data.
+//
+// nullptr DataPtrs can still have a nontrivial device; this allows
+// us to treat zero-size allocations uniformly with non-zero allocations.
+//
+class DataPtr {
+private:
+  detail::UniqueVoidPtr ptr_;
+  Device device_;
+public:
+  // Choice of CPU here is arbitrary; if there's an "undefined" device
+  // we could use that too
+  DataPtr() : ptr_(), device_(kCPU) {}
+  DataPtr(void* data, Device device)
+    : ptr_(data), device_(device) {}
+  DataPtr(void* data, void* ctx, DeleterFnPtr ctx_deleter, Device device)
+    : ptr_(data, ctx, ctx_deleter), device_(device) {}
+  void* operator->() const { return ptr_.get(); }
+  void* get() const { return ptr_.get(); }
+  void* get_context() const { return ptr_.get_context(); }
+  void* release_context() { return ptr_.release_context(); }
+  operator bool() const { return static_cast<bool>(ptr_); }
+  template <typename T>
+  T* cast_context(DeleterFnPtr expected_deleter) const {
+    return ptr_.cast_context<T>(expected_deleter);
+  }
+  Device device() const { return device_; }
+};
+
+// NB: Device is NOT tested for here; a CUDA nullptr is as much a nullptr as a
+// CPU nullptr
+
+inline bool operator==(const at::DataPtr& dp, std::nullptr_t) noexcept { return !dp; }
+inline bool operator==(std::nullptr_t, const at::DataPtr& dp) noexcept { return !dp; }
+inline bool operator!=(const at::DataPtr& dp, std::nullptr_t) noexcept { return dp; }
+inline bool operator!=(std::nullptr_t, const at::DataPtr& dp) noexcept { return dp; }
+
+// Note [raw_allocate/raw_deallocate and Thrust]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// Thrust's support for custom allocators requires us to write something
+// like this:
+//
+//  class ThrustAllocator {
+//    char* allocate(size_t);
+//    void deallocate(char*, size_t);
+//  };
+//
+// This is not good for our unique_ptr based allocator interface, as
+// there is no way to get to the context when we free.
+//
+// However, in some cases the context is exactly the same as
+// the data pointer.  In this case, we can support the "raw"
+// allocate and deallocate interface.  This is what
+// raw_deleter signifies.  By default, it returns a nullptr, which means that
+// the raw interface is not implemented.  Be sure to implement it whenever
```

Comment:
nit: this last sentence isn't actionable -- what happens if I don't do it?

Reply:
new text

```
-// possible.
+// possible, or the raw interface will incorrectly reported as unsupported,
+// when it is actually possible.
```